tidybulk - part of tidyTranscriptomics
================

<!-- badges: start -->

[![Lifecycle:maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)

<!-- badges: end -->

# HPCell

HPCell is a workflow framework designed for use with single-cell RNA
sequencing studies (scRNA-seq), which helps in processing and analyzing
the vast amount of data generated by these studies. It is built on the
Targets framework within the R programming language, offering an
approachable solution for those already proficient in R. The key
features of HPCell include:

- **Native R Pipeline:** HPCell is developed to work natively within the
  R environment, enhancing usability for R users without requiring them
  to learn new, workflow-specific languages.

- **High Performance Computing Support:** HPCell supports scaling for
  large datasets and enables parallel processing of tasks on High
  Performance Computing platforms.

- **Reproducibility and Consistency:** The framework ensures
  reproducibility and consistent execution environments through
  automatic dependency generation.

## Install HPCell

``` r
remote::install_github("stemangiola/HPCell")
```

``` r
library(HPCell)
library(tidybulk)

# To keep the original non-tidy SummarizedExperiment view
options("restore_SummarizedExperiment_show" = TRUE)
```

## High-performance single-cell, pseudobulk random-effect modelling

### One datasets/cell-type

The interface is intuitive and consistent with `tidybulk`
`test_differential_abundance()`. By default `HPCell` uses
`test_differential_abundance(..., method = "glmmSeq_lme4")`

``` r
# Setup your job submission system
slurm = crew.cluster::crew_controller_slurm(
  name = "slurm",
  slurm_memory_gigabytes_per_cpu = 5,
  slurm_cpus_per_task = 1,
  workers = 200,
  verbose = T
)

# Perform the analyses. 
tidySummarizedExperiment::se |>
  keep_abundant(factor_of_interest = dex) |>
  
  # Spread the workload onto 200 workers and collects the results seamlessly
  test_differential_abundance_hpc(
    ~ dex + (1 | cell), 
    computing_resources = slurm
  )
```

    ## â–¶ start target file_data
    ## â— built target file_data [21.311 seconds]
    ## â–¶ start target abundance
    ## â— built target abundance [0 seconds]
    ## â–¶ start target file_formula
    ## â— built target file_formula [0.001 seconds]
    ## â–¶ start target pseudobulk_df_tissue
    ## â— built target pseudobulk_df_tissue [0.016 seconds]
    ## â–¶ start branch pseudobulk_df_tissue_dispersion_5a2c6e14
    ## Submitted batch job 14937770
    ## â–¶ start target number_of_workers
    ## â— built target number_of_workers [0.004 seconds]
    ## â–¶ start target number_of_datasets
    ## â— built target number_of_datasets [0.001 seconds]
    ## â— built branch pseudobulk_df_tissue_dispersion_5a2c6e14 [0.796 seconds]
    ## â— built pattern pseudobulk_df_tissue_dispersion
    ## â–¶ start branch pseudobulk_df_tissue_split_by_gene_64917c93
    ## â— built branch pseudobulk_df_tissue_split_by_gene_64917c93 [0.103 seconds]
    ## â— built pattern pseudobulk_df_tissue_split_by_gene
    ## â–¶ start target pseudobulk_df_tissue_split_by_gene_grouped
    ## â— built target pseudobulk_df_tissue_split_by_gene_grouped [0.075 seconds]
    ## â–¶ start branch estimates_chunk_f60975ce
    ## â— built branch estimates_chunk_f60975ce [23.002 seconds]
    ## â— built pattern estimates_chunk
    ## â–¶ end pipeline [1.155 minutes]
    ## Warning messages:
    ## 1: replacing previous import â€˜tidySingleCellExperiment::plot_lyâ€™ by â€˜tidySummarizedExperiment::plot_lyâ€™ when loading â€˜HPCellâ€™ 
    ## 2: replacing previous import â€˜tidySingleCellExperiment::tidyâ€™ by â€˜tidySummarizedExperiment::tidyâ€™ when loading â€˜HPCellâ€™ 
    ## 3: replacing previous import â€˜tidySummarizedExperiment::tidyâ€™ by â€˜tidyseurat::tidyâ€™ when loading â€˜HPCellâ€™ 
    ## 4: replacing previous import â€˜tidySummarizedExperiment::plot_lyâ€™ by â€˜tidyseurat::plot_lyâ€™ when loading â€˜HPCellâ€™ 
    ## 5: replacing previous import â€˜tidySingleCellExperiment::join_transcriptsâ€™ by â€˜tidyseurat::join_transcriptsâ€™ when loading â€˜HPCellâ€™ 
    ## 6: 2 targets produced warnings. Run targets::tar_meta(fields = warnings, complete_only = TRUE) for the messages.

    ## HPCell says: Start collecting results.

    ## class: RangedSummarizedExperiment 
    ## dim: 25 8 
    ## metadata(1): ''
    ## assays(1): counts
    ## rownames(25): ENSG00000122707 ENSG00000104979 ... ENSG00000246982
    ##   ENSG00000115687
    ## rowData names(26): bla .abundant ... P_dex P_dex_adjusted
    ## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521
    ## colData names(9): SampleName cell ... Sample BioSample

### Many datasets/cell-types

Sometime we do pseudobulk analyses for each cell type. HPCell allows to
scale all those, in a coherent paralleisation to the HPC.

This would be the input dataset

``` r
# A tibble: 22 Ã— 3
   cell_type_harmonised data            formula  
   <chr>                <list>          <list>   
 1 b memory             <SmmrzdEx[,35]> <formula>
 2 b naive              <SmmrzdEx[,35]> <formula>
 3 plasma               <SmmrzdEx[,35]> <formula>
 4 ilc                  <SmmrzdEx[,38]> <formula>
 5 cd4 th1              <SmmrzdEx[,37]> <formula>
 6 cd4 th2              <SmmrzdEx[,40]> <formula>
 7 mait                 <SmmrzdEx[,26]> <formula>
 8 cd8 naive            <SmmrzdEx[,28]> <formula>
 9 cd8 tcm              <SmmrzdEx[,36]> <formula>
10 macrophage           <SmmrzdEx[,21]> <formula>
# â„¹ 12 more rows
# â„¹ Use `print(n = ...)` to see more rows
```

And here the call to the `map` version of
`test_differential_abundance_hpc`

``` r
nested_se |>
      mutate(data = map2_test_differential_abundance_hpc(
        data,
        formula ,
        computing_resources = slurm
      ))
```

This is the output

``` r
# A tibble: 22 Ã— 3
   cell_type_harmonised data            formula  
   <chr>                <list>          <list>   
 1 b memory             <SmmrzdEx[,35]> <formula>
 2 b naive              <SmmrzdEx[,35]> <formula>
 3 plasma               <SmmrzdEx[,35]> <formula>
 4 ilc                  <SmmrzdEx[,38]> <formula>
 5 cd4 th1              <SmmrzdEx[,37]> <formula>
 6 cd4 th2              <SmmrzdEx[,40]> <formula>
 7 mait                 <SmmrzdEx[,26]> <formula>
 8 cd8 naive            <SmmrzdEx[,28]> <formula>
 9 cd8 tcm              <SmmrzdEx[,36]> <formula>
10 macrophage           <SmmrzdEx[,21]> <formula>
# â„¹ 12 more rows
# â„¹ Use `print(n = ...)` to see more rows
```

## High-performance single-cell preprocessing

Flowchart

<https://app.mural.co/t/covid7029/m/covid7029/1656652076667/c47e104697d76b36b8dee3bd05d8de9d96d99efd?sender=udbe1ea99c9618abb07196978>

### load input data

``` r
# Load input data (can be a list of directories or single directory)
library(Seurat)
library(scRNAseq)
input_data_path =  tempfile(tmpdir = ".") |> paste0(".rds")
HeOrganAtlasData(ensembl=FALSE,location=FALSE)[, 1:400] |>
  as.Seurat(data = NULL) |>
  saveRDS(input_data_path)
```

### Execute Targets workflow and load results

``` r
# Running the pipeline
preprocessed_seurat = run_targets_pipeline(
    input_data = input_data_path,
    tissue = "pbmc",
    filter_empty_droplets = TRUE,
    sample_column = "Tissue"
)
```

    ## Warning: Targets and globals must have unique names. Ignoring global objects
    ## that conflict with target names: sample_column, tissue, filter_empty_droplets.
    ## Warnings like this one are important, but if you must suppress them, you can do
    ## so with Sys.setenv(TAR_WARN = "false").

    ## âœ” skip target reference_file

    ## âœ” skip target tissue_file

    ## âœ” skip target sample_column_file

    ## âœ” skip target sample_column

    ## âœ” skip target tissue

    ## âœ” skip target reference_label_fine

    ## âœ” skip target read_file

    ## âœ” skip branch input_read_58a1fafa

    ## âœ” skip pattern input_read

    ## âœ” skip target file

    ## âœ” skip target filtered_file

    ## âœ” skip target filter_empty_droplets

    ## âœ” skip branch empty_droplets_tbl_2a87e9d4

    ## âœ” skip pattern empty_droplets_tbl

    ## âœ” skip branch cell_cycle_score_tbl_ab819f51

    ## âœ” skip pattern cell_cycle_score_tbl

    ## âœ” skip target reference_read

    ## âœ” skip branch annotation_label_transfer_tbl_ab819f51

    ## âœ” skip pattern annotation_label_transfer_tbl

    ## âœ” skip branch alive_identification_tbl_bfea2a3c

    ## âœ” skip pattern alive_identification_tbl

    ## âœ” skip branch doublet_identification_tbl_c49acd50

    ## âœ” skip pattern doublet_identification_tbl

    ## âœ” skip branch non_batch_variation_removal_S_2405b948

    ## âœ” skip pattern non_batch_variation_removal_S

    ## âœ” skip branch preprocessing_output_S_269c823b

    ## âœ” skip pattern preprocessing_output_S

    ## âœ” skip target reference_label_coarse

    ## âœ” skip target pseudobulk_preprocessing_SE

    ## âœ” skip pipeline [0.227 seconds]

    ## HPCell says: you can read your output executing tar_read(preprocessing_output_S, store = "./")

``` r
# Load results
preprocessed_seurat
```

    ## $preprocessing_output_S_269c823b
    ## # A Seurat-tibble abstraction: 290 Ã— 46
    ## # [90mFeatures=9560 | Cells=290 | Active assay=SCT | Assays=originalexp, SCT[0m
    ##    .cell    orig.ident nCount_originalexp nFeature_originalexp Tissue nCount_RNA
    ##    <chr>    <fct>                   <dbl>                <int> <chr>       <int>
    ##  1 Bladderâ€¦ Bladder                  1133                  592 Bladdâ€¦       1152
    ##  2 Bladderâ€¦ Bladder                  3495                 1374 Bladdâ€¦       3551
    ##  3 Bladderâ€¦ Bladder                  1297                  797 Bladdâ€¦       1599
    ##  4 Bladderâ€¦ Bladder                  2071                  953 Bladdâ€¦       2355
    ##  5 Bladderâ€¦ Bladder                  2166                 1102 Bladdâ€¦       2474
    ##  6 Bladderâ€¦ Bladder                  2486                 1185 Bladdâ€¦       2734
    ##  7 Bladderâ€¦ Bladder                  3175                 1418 Bladdâ€¦       3727
    ##  8 Bladderâ€¦ Bladder                  1847                  932 Bladdâ€¦       2013
    ##  9 Bladderâ€¦ Bladder                  2546                 1104 Bladdâ€¦       2649
    ## 10 Bladderâ€¦ Bladder                   969                  574 Bladdâ€¦       1138
    ## # â„¹ 280 more rows
    ## # â„¹ 40 more variables: nFeature_RNA <int>, percent.mito <dbl>,
    ## #   RNA_snn_res.orig <int>, seurat_clusters <int>,
    ## #   Cell_type_in_each_tissue <chr>, Cell_type_in_merged_data <chr>,
    ## #   reclustered.broad <chr>, reclustered.fine <chr>, Total <int>,
    ## #   LogProb <dbl>, PValue <dbl>, Limited <lgl>, FDR <dbl>, empty_droplet <lgl>,
    ## #   rank <dbl>, total <int>, fitted <dbl>, knee <dbl>, inflection <dbl>, â€¦

### Include reference dataset for azimuth annotation

Details to come.

``` r
input_reference_path  <- "reference_azimuth.rds" 
reference_url <- "https://atlas.fredhutch.org/data/nygc/multimodal/pbmc_multimodal.h5seurat" 
download.file(reference_url, input_reference_path) 
LoadH5Seurat(input_reference_path) |> saveRDS(input_reference_path)
```
