tidybulk - part of tidyTranscriptomics
================

<!-- badges: start -->

[![Lifecycle:maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)

<!-- badges: end -->

# HPCell

HPCell is a workflow framework designed for use with single-cell RNA
sequencing studies (scRNA-seq), which helps in processing and analyzing
the vast amount of data generated by these studies. It is built on the
Targets framework within the R programming language, offering an
approachable solution for those already proficient in R. The key
features of HPCell include:

- **Native R Pipeline:** HPCell is developed to work natively within the
  R environment, enhancing usability for R users without requiring them
  to learn new, workflow-specific languages.

- **High Performance Computing Support:** HPCell supports scaling for
  large datasets and enables parallel processing of tasks on High
  Performance Computing platforms.

- **Reproducibility and Consistency:** The framework ensures
  reproducibility and consistent execution environments through
  automatic dependency generation.

## Install HPCell

``` r
remote::install_github("stemangiola/HPCell")
```

``` r
library(HPCell)
library(tidybulk)

# To keep the original non-tidy SummarizedExperiment view
options("restore_SummarizedExperiment_show" = TRUE)
```

## High-performance single-cell, pseudobulk random-effect modelling

### One datasets/cell-type

The interface is intuitive and consistent with `tidybulk`
`test_differential_abundance()`. By default `HPCell` uses
`test_differential_abundance(..., method = "glmmSeq_lme4")`

``` r
# Setup your job submission system
slurm = crew.cluster::crew_controller_slurm(
  name = "slurm",
  slurm_memory_gigabytes_per_cpu = 5,
  slurm_cpus_per_task = 1,
  workers = 200,
  verbose = T
)

# Perform the analyses. 
tidySummarizedExperiment::se |>
  keep_abundant(factor_of_interest = dex) |>
  
  # Spread the workload onto 200 workers and collects the results seamlessly
  test_differential_abundance_hpc(
    ~ dex + (1 | cell), 
    computing_resources = slurm
  )
```

    ## ▶ start target file_data
    ## ● built target file_data [21.311 seconds]
    ## ▶ start target abundance
    ## ● built target abundance [0 seconds]
    ## ▶ start target file_formula
    ## ● built target file_formula [0.001 seconds]
    ## ▶ start target pseudobulk_df_tissue
    ## ● built target pseudobulk_df_tissue [0.016 seconds]
    ## ▶ start branch pseudobulk_df_tissue_dispersion_5a2c6e14
    ## Submitted batch job 14937770
    ## ▶ start target number_of_workers
    ## ● built target number_of_workers [0.004 seconds]
    ## ▶ start target number_of_datasets
    ## ● built target number_of_datasets [0.001 seconds]
    ## ● built branch pseudobulk_df_tissue_dispersion_5a2c6e14 [0.796 seconds]
    ## ● built pattern pseudobulk_df_tissue_dispersion
    ## ▶ start branch pseudobulk_df_tissue_split_by_gene_64917c93
    ## ● built branch pseudobulk_df_tissue_split_by_gene_64917c93 [0.103 seconds]
    ## ● built pattern pseudobulk_df_tissue_split_by_gene
    ## ▶ start target pseudobulk_df_tissue_split_by_gene_grouped
    ## ● built target pseudobulk_df_tissue_split_by_gene_grouped [0.075 seconds]
    ## ▶ start branch estimates_chunk_f60975ce
    ## ● built branch estimates_chunk_f60975ce [23.002 seconds]
    ## ● built pattern estimates_chunk
    ## ▶ end pipeline [1.155 minutes]
    ## Warning messages:
    ## 1: replacing previous import ‘tidySingleCellExperiment::plot_ly’ by ‘tidySummarizedExperiment::plot_ly’ when loading ‘HPCell’ 
    ## 2: replacing previous import ‘tidySingleCellExperiment::tidy’ by ‘tidySummarizedExperiment::tidy’ when loading ‘HPCell’ 
    ## 3: replacing previous import ‘tidySummarizedExperiment::tidy’ by ‘tidyseurat::tidy’ when loading ‘HPCell’ 
    ## 4: replacing previous import ‘tidySummarizedExperiment::plot_ly’ by ‘tidyseurat::plot_ly’ when loading ‘HPCell’ 
    ## 5: replacing previous import ‘tidySingleCellExperiment::join_transcripts’ by ‘tidyseurat::join_transcripts’ when loading ‘HPCell’ 
    ## 6: 2 targets produced warnings. Run targets::tar_meta(fields = warnings, complete_only = TRUE) for the messages.

    ## HPCell says: Start collecting results.

    ## class: RangedSummarizedExperiment 
    ## dim: 25 8 
    ## metadata(1): ''
    ## assays(1): counts
    ## rownames(25): ENSG00000122707 ENSG00000104979 ... ENSG00000246982
    ##   ENSG00000115687
    ## rowData names(26): bla .abundant ... P_dex P_dex_adjusted
    ## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521
    ## colData names(9): SampleName cell ... Sample BioSample

### Many datasets/cell-types

Sometime we do pseudobulk analyses for each cell type. HPCell allows to
scale all those, in a coherent paralleisation to the HPC.

This would be the input dataset

``` r
# A tibble: 22 × 3
   cell_type_harmonised data            formula  
   <chr>                <list>          <list>   
 1 b memory             <SmmrzdEx[,35]> <formula>
 2 b naive              <SmmrzdEx[,35]> <formula>
 3 plasma               <SmmrzdEx[,35]> <formula>
 4 ilc                  <SmmrzdEx[,38]> <formula>
 5 cd4 th1              <SmmrzdEx[,37]> <formula>
 6 cd4 th2              <SmmrzdEx[,40]> <formula>
 7 mait                 <SmmrzdEx[,26]> <formula>
 8 cd8 naive            <SmmrzdEx[,28]> <formula>
 9 cd8 tcm              <SmmrzdEx[,36]> <formula>
10 macrophage           <SmmrzdEx[,21]> <formula>
# ℹ 12 more rows
# ℹ Use `print(n = ...)` to see more rows
```

And here the call to the `map` version of
`test_differential_abundance_hpc`

``` r
nested_se |>
      mutate(data = map2_test_differential_abundance_hpc(
        data,
        formula ,
        computing_resources = slurm
      ))
```

This is the output

``` r
# A tibble: 22 × 3
   cell_type_harmonised data            formula  
   <chr>                <list>          <list>   
 1 b memory             <SmmrzdEx[,35]> <formula>
 2 b naive              <SmmrzdEx[,35]> <formula>
 3 plasma               <SmmrzdEx[,35]> <formula>
 4 ilc                  <SmmrzdEx[,38]> <formula>
 5 cd4 th1              <SmmrzdEx[,37]> <formula>
 6 cd4 th2              <SmmrzdEx[,40]> <formula>
 7 mait                 <SmmrzdEx[,26]> <formula>
 8 cd8 naive            <SmmrzdEx[,28]> <formula>
 9 cd8 tcm              <SmmrzdEx[,36]> <formula>
10 macrophage           <SmmrzdEx[,21]> <formula>
# ℹ 12 more rows
# ℹ Use `print(n = ...)` to see more rows
```

## High-performance single-cell preprocessing

Flowchart

<https://app.mural.co/t/covid7029/m/covid7029/1656652076667/c47e104697d76b36b8dee3bd05d8de9d96d99efd?sender=udbe1ea99c9618abb07196978>

### load input data

``` r
# Load input data (can be a list of directories or single directory)
library(Seurat)
<<<<<<< HEAD
```

    ## Attaching SeuratObject

    ## 'SeuratObject' was built under R 4.3.0 but the current version is
    ## 4.3.1; it is recomended that you reinstall 'SeuratObject' as the ABI
    ## for R may have changed

``` r
library(scRNAseq)
```

    ## Loading required package: SingleCellExperiment

    ## Loading required package: SummarizedExperiment

    ## Loading required package: MatrixGenerics

    ## Loading required package: matrixStats

    ## 
    ## Attaching package: 'matrixStats'

    ## The following objects are masked from 'package:Biobase':
    ## 
    ##     anyMissing, rowMedians

    ## 
    ## Attaching package: 'MatrixGenerics'

    ## The following objects are masked from 'package:matrixStats':
    ## 
    ##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    ##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    ##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    ##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    ##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    ##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    ##     colWeightedMeans, colWeightedMedians, colWeightedSds,
    ##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    ##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    ##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    ##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    ##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    ##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    ##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    ##     rowWeightedSds, rowWeightedVars

    ## The following object is masked from 'package:Biobase':
    ## 
    ##     rowMedians

    ## Loading required package: GenomicRanges

    ## Loading required package: stats4

    ## Loading required package: S4Vectors

    ## 
    ## Attaching package: 'S4Vectors'

    ## The following object is masked from 'package:utils':
    ## 
    ##     findMatches

    ## The following objects are masked from 'package:base':
    ## 
    ##     expand.grid, I, unname

    ## Loading required package: IRanges

    ## Loading required package: GenomeInfoDb

    ## 
    ## Attaching package: 'SummarizedExperiment'

    ## The following object is masked from 'package:SeuratObject':
    ## 
    ##     Assays

    ## The following object is masked from 'package:Seurat':
    ## 
    ##     Assays

``` r
input_data_path =  tempfile(tmpdir = "~") |> paste0(".rds")
HeOrganAtlasData(ensembl=FALSE,location=FALSE)[, 1:400] |> 
  as.Seurat(data = NULL) |> 
  saveRDS(input_data_path) 
```

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

    ## see ?scRNAseq and browseVignettes('scRNAseq') for documentation

    ## loading from cache

library(scRNAseq)
input_data_path =  tempfile(tmpdir = ".") |> paste0(".rds")
HeOrganAtlasData(ensembl=FALSE,location=FALSE)[, 1:400] |>
  as.Seurat(data = NULL) |>
  saveRDS(input_data_path)
```

### Execute Targets workflow and load results
=======
Execute Targets workflow and load results
    ## ▶ start target sample_column_file

    ## ● built target sample_column_file [0.001 seconds]

    ## ▶ start target sample_column

    ## ● built target sample_column [0.001 seconds]

    ## ▶ start target filtered_file

    ## ● built target filtered_file [0.001 seconds]

    ## ▶ start target filter_input

    ## ● built target filter_input [0.001 seconds]

    ## ▶ start target file

    ## ● built target file [0.001 seconds]

    ## ▶ start target read_file

    ## ● built target read_file [0.001 seconds]

    ## ▶ start branch input_read_afac7453

    ## ● built branch input_read_afac7453 [0.046 seconds]

    ## ● built pattern input_read

    ## ▶ start branch input_read_RNA_assay_639a47c9

    ## ● built branch input_read_RNA_assay_639a47c9 [0.123 seconds]

    ## ● built pattern input_read_RNA_assay

    ## ▶ start branch empty_droplets_tbl_17fcd675

    ## ● built branch empty_droplets_tbl_17fcd675 [16.483 seconds]

    ## ● built pattern empty_droplets_tbl

    ## ▶ start branch cell_cycle_score_tbl_87491175

    ## ● built branch cell_cycle_score_tbl_87491175 [0.256 seconds]

    ## ● built pattern cell_cycle_score_tbl

    ## ▶ start branch annotation_label_transfer_tbl_87491175

    ## ● built branch annotation_label_transfer_tbl_87491175 [23.661 seconds]

    ## ● built pattern annotation_label_transfer_tbl

    ## ▶ start branch alive_identification_tbl_9ec1e7c5

    ## ● built branch alive_identification_tbl_9ec1e7c5 [1.604 seconds]

    ## ● built pattern alive_identification_tbl

    ## ▶ start branch non_batch_variation_removal_S_532e226b

    ## ● built branch non_batch_variation_removal_S_532e226b [8.257 seconds]

    ## ● built pattern non_batch_variation_removal_S

    ## ▶ start branch doublet_identification_tbl_0883335c

    ## ● built branch doublet_identification_tbl_0883335c [5.521 seconds]

    ## ● built pattern doublet_identification_tbl

    ## ▶ start branch preprocessing_output_S_42aad64a

    ## ● built branch preprocessing_output_S_42aad64a [0.258 seconds]

    ## ● built pattern preprocessing_output_S

    ## ▶ start target pseudobulk_preprocessing_SE

    ## ● built target pseudobulk_preprocessing_SE [4.882 seconds]

    ## ▶ end pipeline [1.81 minutes]

    ## Warning: 4 targets produced warnings. Run targets::tar_meta(fields = warnings,
    ## complete_only = TRUE) for the messages.
 
    ## HPCell says: you can read your output executing tar_read(preprocessing_output_S, store = "./")

``` r
# Load results
preprocessed_seurat
```
    ## $preprocessing_output_S_269c823b
    ## # A Seurat-tibble abstraction: 290 × 46
    ## # [90mFeatures=9560 | Cells=290 | Active assay=SCT | Assays=originalexp, SCT[0m

    ##    .cell    orig.ident nCount_originalexp nFeature_originalexp Tissue nCount_RNA
    ##    <chr>    <fct>                   <dbl>                <int> <chr>       <int>
    ##  1 Bladder… Bladder                  1133                  592 Bladd…       1152
    ##  2 Bladder… Bladder                  3495                 1374 Bladd…       3551
    ##  3 Bladder… Bladder                  1297                  797 Bladd…       1599
    ##  4 Bladder… Bladder                  2071                  953 Bladd…       2355
    ##  5 Bladder… Bladder                  2166                 1102 Bladd…       2474
    ##  6 Bladder… Bladder                  2486                 1185 Bladd…       2734
    ##  7 Bladder… Bladder                  3175                 1418 Bladd…       3727
    ##  8 Bladder… Bladder                  1847                  932 Bladd…       2013
    ##  9 Bladder… Bladder                  2546                 1104 Bladd…       2649
    ## 10 Bladder… Bladder                   969                  574 Bladd…       1138
    ## # ℹ 280 more rows
    ## # ℹ 40 more variables: nFeature_RNA <int>, percent.mito <dbl>,
    ## #   RNA_snn_res.orig <int>, seurat_clusters <int>,
    ## #   Cell_type_in_each_tissue <chr>, Cell_type_in_merged_data <chr>,
    ## #   reclustered.broad <chr>, reclustered.fine <chr>, Total <int>,
    ## #   LogProb <dbl>, PValue <dbl>, Limited <lgl>, FDR <dbl>, empty_droplet <lgl>,
    ## #   rank <dbl>, total <int>, fitted <dbl>, knee <dbl>, inflection <dbl>, …

### Include reference dataset for azimuth annotation

Details to come.

``` r
input_reference_path  <- "reference_azimuth.rds" 
reference_url <- "https://atlas.fredhutch.org/data/nygc/multimodal/pbmc_multimodal.h5seurat" 
download.file(reference_url, input_reference_path) 
LoadH5Seurat(input_reference_path) |> saveRDS(input_reference_path)
```
# HPCell Pipeline Documentation

## Overview
The `run_targets_pipeline` function orchestrates a targeted workflow for processing single-cell RNA sequencing data, using the `targets` R package. This documentation explains the pipeline's structure, methods used, and provides code snippets to illustrate its operation. It's designed to help beginners understand both the how and the why of each step.

## Introduction to `targets`
The `targets` package in R automates and streamlines the execution of complex data analysis pipelines. It manages dependencies between steps (targets) and only reruns parts of the workflow that have changed. Users can also map targets dynamically to manage a list of input and/or parameters. 

## Pipeline Steps and Methods

``` r
run_targets_pipeline <- function(
    input_data, 
    store = "./", 
    input_reference = NULL,
    tissue,
    computing_resources = crew_controller_local(workers = 1), 
    debug_step = NULL,
    filter_empty_droplets = TRUE, 
    RNA_assay_name = "RNA", 
    sample_column = "sample"
) {
  sample_column = enquo(sample_column)
  
  # Saving Input Parameters
  # User-defined parameters and data are saved as .rds files. This step ensures that all 
  # necessary information is available to the targets pipeline as input targets.
  
  input_data |> saveRDS("input_file.rds")
  input_reference |> saveRDS("input_reference.rds")
  tissue |> saveRDS("tissue.rds")
  computing_resources |> saveRDS("temp_computing_resources.rds")
  filter_empty_droplets |> saveRDS("filter_empty_droplets.rds")
  sample_column |> saveRDS("sample_column.rds")
  
  # Configuring'targets' 
  # Using tar_option_set, we configure global settings for the targets pipeline, specifying 
  # required packages, memory options, and computing resources.
  
  tar_script({
    
    library(targets)
    library(tarchetypes)
    library(crew)
    library(crew.cluster)
    
    computing_resources = readRDS("temp_computing_resources.rds")
    #-----------------------#
    # Packages
    #-----------------------#
    tar_option_set(
      packages = c(
        "HPCell",
        "readr",
        "dplyr",
        "tidyr",
        "ggplot2",
        "purrr",
        "Seurat",
        "tidyseurat",
        "glue",
        "scater",
        "DropletUtils",
        "EnsDb.Hsapiens.v86",
        "here",
        "stringr",
        "readr",
        "rlang",
        "scuttle",
        "scDblFinder",
        "ggupset",
        "tidySummarizedExperiment",
        "broom",
        "tarchetypes",
        "SeuratObject",
        "SingleCellExperiment", 
        "SingleR", 
        "celldex", 
        "tidySingleCellExperiment", 
        "tibble", 
        "magrittr",
        "qs", 
        "S4Vectors"
      ),
      memory = "transient",
      garbage_collection = TRUE,
      #trust_object_timestamps = TRUE,
      storage = "worker", 
      retrieval = "worker", 
      #error = "continue",         
      format = "qs", 
      debug = debug_step, # Set the target you want to debug.
      # cue = tar_cue(mode = "never") # Force skip non-debugging outdated targets.
      controller = computing_resources
    )
    
    # Defining targets 
    # Targets represent each step in the data analysis process. 
    # For instance, filtering empty droplets, annotating cell types, and identifying cell cycles 
    # are defined as individual targets within the pipeline.
    
    # Managing dependencies 
    # Each analysis step depends on the output of previous steps. targets tracks these dependencies automatically. 
    
    # Mapping
    # Target are defined that operates over a list of datasets or a range of parameters. 
    # Allows the application the same processing steps to multiple samples or datasets.
    
    target_list = list(
      tar_target(file, "input_file.rds", format = "rds"), 
      tar_target(read_file, readRDS("input_file.rds")),
      #tar_target(reference_file, "input_reference.rds", format = "rds"), 
      tar_target(reference_file, readRDS("input_reference.rds")), 
      tar_target(tissue_file, readRDS("tissue.rds")), 
      tar_target(filtered_file, readRDS("filter_empty_droplets.rds")), 
      tar_target(sample_column_file, readRDS("sample_column.rds")))
    
    #-----------------------#
    # Pipeline
    #-----------------------#
    target_list|> c(list(
      
      # Define input files
      # tarchetypes::tar_files(name= input_track, 
      #                        read_file, 
      #                        deployment = "main"),
      # tarchetypes::tar_files(name= reference_track,
      #                        read_reference_file, 
      #                        deployment = "main"),
      tar_target(filter_empty_droplets, filtered_file, deployment = "main"),
      tar_target(tissue, tissue_file, deployment = "main"),
      tar_target(sample_column, sample_column_file, deployment = "main"),
      tar_target(reference_label_coarse, reference_label_coarse_id(tissue), deployment = "main"), 
      tar_target(reference_label_fine, reference_label_fine_id(tissue), deployment = "main"), 
      # Reading input files
      tar_target(input_read, readRDS(read_file),
                 pattern = map(read_file),
                 iteration = "list", deployment = "main"),

      tar_target(reference_read, reference_file, deployment = "main"),
      
      # Identifying empty droplets
      tar_target(empty_droplets_tbl,
                 empty_droplet_id(input_read, filter_empty_droplets),
                 pattern = map(input_read),
                 iteration = "list"),
      
      # Cell cycle scoring
      tar_target(cell_cycle_score_tbl, cell_cycle_scoring(input_read,
                                                          empty_droplets_tbl),
                 pattern = map(input_read,
                               empty_droplets_tbl),
                 iteration = "list"),
      
      # Annotation label transfer
      tar_target(annotation_label_transfer_tbl,
                 annotation_label_transfer(input_read,
                                           empty_droplets_tbl,
                                           reference_read),
                 pattern = map(input_read,
                               empty_droplets_tbl),
                 iteration = "list"),
      
      # Alive identification
      tar_target(alive_identification_tbl, alive_identification(input_read,
                                                                empty_droplets_tbl,
                                                                annotation_label_transfer_tbl),
                 pattern = map(input_read,
                               empty_droplets_tbl,
                               annotation_label_transfer_tbl),
                 iteration = "list"),
      
      # Doublet identification
      tar_target(doublet_identification_tbl, doublet_identification(input_read,
                                                                    empty_droplets_tbl,
                                                                    alive_identification_tbl,
                                                                    annotation_label_transfer_tbl,
                                                                    reference_label_fine),
                 pattern = map(input_read,
                               empty_droplets_tbl,
                               alive_identification_tbl,
                               annotation_label_transfer_tbl),
                 iteration = "list"),
      
      # Non-batch variation removal
      tar_target(non_batch_variation_removal_S, non_batch_variation_removal(input_read,
                                                                            empty_droplets_tbl,
                                                                            alive_identification_tbl,
                                                                            cell_cycle_score_tbl),
                 pattern = map(input_read,
                               empty_droplets_tbl,
                               alive_identification_tbl,
                               cell_cycle_score_tbl),
                 iteration = "list"),
      
      # Pre-processing output
      tar_target(preprocessing_output_S, preprocessing_output(tissue,
                                                              non_batch_variation_removal_S,
                                                              alive_identification_tbl,
                                                              cell_cycle_score_tbl,
                                                              annotation_label_transfer_tbl,
                                                              doublet_identification_tbl),
                 pattern = map(non_batch_variation_removal_S,
                               alive_identification_tbl,
                               cell_cycle_score_tbl,
                               annotation_label_transfer_tbl,
                               doublet_identification_tbl),
                 iteration = "list"),
      
      # pseudobulk preprocessing for each sample 
      tar_target(create_pseudobulk_sample, create_pseudobulk(preprocessing_output_S, 
                                                                   assays = "RNA", 
                                                                   x = c(Tissue, Cell_type_in_each_tissue)), 
                 pattern = map(preprocessing_output_S), 
                 iteration = "list"),
      
      tar_target(pseudobulk_merge_all_samples, pseudobulk_merge(create_pseudobulk_sample, 
                                                                assays = "RNA", 
                                                                x = c(Tissue)), 
                 pattern = map(create_pseudobulk_sample), 
                 iteration = "list"),
      
      tar_target(calc_UMAP_dbl_report, calc_UMAP(input_read), 
                 pattern = map(input_read), 
                 iteration = "list")
      ))
  }, script = glue("{store}.R"), ask = FALSE)
  
  # Executing the Pipeline
  # The tar_make() function initiates the execution of the pipeline, 
  # processing each target according to the dependencies defined.
  
  tar_make(
    script = glue("{store}.R"),
    store = store, 
    callr_function = NULL
  )
  
  message(glue("HPCell says: you can read your output executing tar_read(preprocessing_output_S, store = \"{store}\")"))
  
  # Accessing Results
  # After the pipeline execution, results can be accessed using tar_read()
  
  tar_read(preprocessing_output_S, store = store)
}
```




