---
title: "Empty droplet report"
author: "SS"
date: "2023-12-07"
output: html_document
params: 
  x1: "NA"
  x2: "NA"
  x3: "NA"
  x4: "NA"
  x5: "NA"
---

```{r,  warning=FALSE, message=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(HPCell)
library(readr)
library(dplyr)
library(tidyr)
library(ggplot2)
library(purrr)
library(Seurat)
library(tidyseurat)
library(glue)
library(scater)
library(DropletUtils)
library(EnsDb.Hsapiens.v86)
library(here)
library(stringr)
library(rlang)
library(scuttle)
library(scDblFinder)
library(ggupset)
library(tidySummarizedExperiment)
library(broom)
library(tarchetypes)
library(SeuratObject)
library(SingleCellExperiment)
library(SingleR)
library(celldex)
library(tidySingleCellExperiment)
library(tibble)
library(magrittr)
library(qs)
library(S4Vectors)

# Subsetting tissues in input data 
# unique_tissues <- unique(input_seurat_abc@meta.data$Tissue)

assay = params$x1[[1]]@assays |> names() |> extract2(1)
# Subset 2 tissues (sample types)
# heart <- subset(input_seurat, subset = Tissue == "Heart")
# trachea <- subset(input_seurat, subset = Tissue == "Trachea")
# params$x1 <- c(heart, trachea)

# # ma plot: mito is green, ribo is red
# col <- rep('black',ncol(empty_droplets_tbl))
# col[rownames(empty_droplets_tbl) %in% mito_genes] <-'green'
# col[rownames(empty_droplets_tbl) %in% ribo_genes] <-'red'
# 
# Process input data
# process_input <- function(input_seurat) {
#   #browser()
#   input<- input_seurat@meta.data |>
#     tibble::rownames_to_column(var = '.cell')
#   
#   # grep('^MT-', rownames(input_seurat[['RNA']]), value=T)
#   #define mito and ribo genes and add the plot:
#   mito_genes <- grep('^MT-', rownames(input_seurat[[assay]]), value=T)
#   ribo_genes <-grep('^RP(S|L)', rownames(input_seurat[[assay]]), value=T)
#   
#   col <- rep('cornflowerblue',ncol(input_seurat))
#   col[rownames(input_seurat) %in% mito_genes] <-'green'
#   col[rownames(input_seurat) %in% ribo_genes] <-'red'
#   col[rownames(input_seurat) %in% NA] <-'grey'
#   sample_numbers = 1:length(list(input_seurat))
#   #sample_names <- unique(input$Tissue)
# }
# processed_input_list <- map(params$x1, process_input)

# input <- input_seurat@meta.data |> 
#     tibble::rownames_to_column(var = '.cell')
#   
# joined_data <- empty_droplets_tbl |>
#     left_join(input |> dplyr::select(.cell, Tissue), by = '.cell')
  
# Defining Tissue names
# sample_names <- sapply(1:length(params$x1), function(i) {
#   #browser()
#         return(params$x1[[i]][[i]])
# })
# 

# sample_names <- lapply(params$x1, function(seurat_obj) {
#     unique(seurat_obj$Tissue)
# })
# sample_names<- unlist(sample_names)
# 
# sample_names <- sapply(1:length(params$x1), function(i) {
#         return(params$x1[[i]][[1]][[1]])
# })
# sample_names
process_input<- function(input_seurat) {
  input<- input_seurat@meta.data |>
  tibble::rownames_to_column(var = '.cell')
  return(input)
}
processed_input_list <- map(params$x1, process_input)
  
# Defining Tissue names
# sample_names <- lapply(params$x1, function(seurat_obj) {
#     seurat_obj |> pull(params$x5)
# })
sample_names<- params$x4
# sample_names<- unlist(sample_names)
```

## Barcode rank plot
```{r echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Process empty droplets data 
empty_df <- function(input_seurat, empty_droplets_tbl) {
  input <- input_seurat@meta.data |> 
    tibble::rownames_to_column(var = '.cell')
  
  joined_data <- empty_droplets_tbl |> 
    left_join(input |> dplyr::select(.cell, params$x5), by = '.cell')

  # Create a data frame with plotting information
  plot_data <- data.frame(
    x = joined_data$rank,
    y = joined_data$total, 
    rank = joined_data$rank, 
    inflection = joined_data$inflection,
    knee = joined_data$knee,
    fitted = joined_data$fitted, 
    empty = joined_data$empty_droplet, 
    FDR = joined_data$FDR,
    Total = joined_data$Total, 
    PValue = joined_data$PValue
  )
  return(plot_data)
}


process_empty_droplet_list <- purrr::map2(params$x1, params$x2, empty_df)

# Combined tibble with an identifier for each tissue/sample
combined_df <- bind_rows(process_empty_droplet_list, .id = "tissue_id") %>%
  mutate(tissue_id = factor(tissue_id, labels = params$x4))

# Generate plot
plot <- ggplot(combined_df, aes(x = x, y = y)) +
    geom_point(color = 'lightblue', alpha = 0.5) +
    scale_x_log10() +
    scale_y_log10() +
    geom_line(aes(x = rank, y = fitted), color='darkblue') +
    geom_hline(aes(yintercept = knee), color='red') +
    geom_hline(aes(yintercept = inflection), color='forestgreen') +
    scale_linetype_manual(values = c("knee" = "dashed", "inflection" = "dashed"),
                          guide = guide_legend(override.aes = list(color = c("forestgreen", "red")))
                          ) +
    facet_wrap(~tissue_id, scales = "free") +
    theme_minimal() +
    labs(x = "Barcodes", y = "Total UMI count", color = "Legend") +
    theme(legend.position = "bottom")  # Adjust legend position as needed

print(plot)
```

## Proportion of empty droplets
```{r,  warning=FALSE, message=FALSE, echo=FALSE}
empty_count <- function(df) {
  # Count the TRUE and FALSE values in the empty_droplet column
  tibble <- df %>%
    group_by(tissue_id) %>%
    summarise(
      Empty_count = sum(empty == TRUE),
      Cell_count = sum(empty == FALSE)
    )
  return(tibble)
}

# Apply the function to the combined_df
empty_count_results <- empty_count(combined_df)
empty_count_results
```

## Number and proportion of cells (non-empty droplets), everything above knee is retained.
```{r,  warning=FALSE, message=FALSE, echo=FALSE}
# Number of non-empty droplets -------------------------------------------------
empty_table <- function(df) {
  # Count the TRUE and FALSE values in the empty_droplet column
  tibble <- df %>%
    group_by(tissue_id) %>%
    summarise(
     "Number: True cells (FDR<0.001)" = sum(FDR < 0.001, na.rm = TRUE),  # Count of FDR values less than 0.001
     "Proportion: True cells (FDR<0.001)" = mean(FDR < 0.001, na.rm = TRUE)  # Proportion of FDR values less than 0.001
    )
  return(tibble)
}
empty_count_results <- empty_table(combined_df)
empty_count_results
```

<!-- ## Count of cells vs empty droplets -->
<!-- ```{r,  warning=FALSE, message=FALSE, echo=FALSE} -->
<!-- count <- function(df) { -->
<!--   is.cell <- df$FDR <= 0.001 -->
<!--   tibble<- df %>% -->
<!--   group_by(tissue_id) %>% -->
<!--   summarise( -->
<!--     Cells = sum(is.cell, na.rm = TRUE),      # Count of TRUE values, NA values removed -->
<!--     Empty_droplets = sum(!is.cell, na.rm = TRUE)   # Count of FALSE values, NA values removed -->
<!--   ) -->
<!--   return(tibble) -->
<!-- } -->
<!-- count_results <- count(combined_df) -->
<!-- count_results -->
<!-- ``` -->


## Histogram of p-values: (only if empty droplets have been identified)

- Shows the distribution of p-values for droplets in the lower 10 percentile of total within each tissue 
- A low p-value signifies significance therefore we would reject those droplets as empty
```{r,  warning=FALSE, message=FALSE, echo=FALSE}
hist_p_val <- function(df) {
  if(df |> dplyr::filter(empty) |> nrow() != 0){
  df_filtered <- df %>%
  group_by(tissue_id) %>%
  dplyr::filter(empty) %>%
  mutate(Total_quantile = quantile(Total[Total > 0], 0.1)) %>%
  dplyr::filter(Total <= Total_quantile & Total > 0) %>%
  ungroup()

plot_hist <- ggplot(df_filtered, aes(x = PValue)) +
  geom_histogram(binwidth = 0.2, fill = "cornflowerblue", color = "grey") +
  facet_wrap(~ tissue_id) +
  labs(x = "P-value", y = "Frequency") +
  ggtitle("Droplets with 0 < libsize <= 10th Percentile of Total per Tissue") + 
  theme_minimal()
}}

plot_hist <- hist_p_val(combined_df)
plot_hist
```

## Percentage of reads assigned to mitochondrial transcrips against library size

Scatter plot comparing mitochondrial content percentage to total count of RNA sequencing reads across different samples (in this case tissues)

The X-axis is on a logarithmic scale and represents the total count of RNA sequencing reads per cell, while the Y-axis shows the percentage of those reads that are mitochondrial. Each point on the plot represents a single cell.

```{r,  warning=FALSE, message=FALSE, echo=FALSE}
plot_mito_data <- function(input_seurat, tissue_name, annotation_labels){
  #browser()
  rna_counts <- GetAssayData(input_seurat, layer = "counts", assay=assay)
  which_mito = rownames(input_seurat) |> str_which("^MT")
  # Compute per-cell QC metrics
  qc_metrics <- perCellQCMetrics(rna_counts, subsets=list(Mito=which_mito)) %>%
    as_tibble(rownames = ".cell") %>%
    dplyr::select(-sum, -detected)

  #Identify mitochondrial content
  # mitochondrion <- qc_metrics %>%
  #   left_join(annotation_labels, by = ".cell") %>%
  #   nest(data = -blueprint_first.labels.fine) %>%
  #   mutate(data = map(data, ~ .x %>%
  #                       mutate(high_mitochondrion = isOutlier(subsets_Mito_percent, type="higher"),
  #                              high_mitochondrion = as.logical(high_mitochondrion)))) %>%
  #   unnest(cols = data)
  mitochondrion <- qc_metrics %>%
    left_join(annotation_labels, by = ".cell") %>%
    mutate(high_mitochondrion = isOutlier(subsets_Mito_percent, type="higher")) %>%
    mutate(high_mitochondrion = as.logical(high_mitochondrion), 
           tissue_name = tissue_name) %>%
    group_by(tissue_name) %>%
    mutate(
      discard = isOutlier(subsets_Mito_percent, type = "higher"),
      threshold = attr(discard, "threshold")["higher"]
    ) %>%
    ungroup()

  # discard <- isOutlier(mitochondrion$subsets_Mito_percent, type = "higher")
  # threshold <- attr(discard, "threshold")["higher"]
  plot_mito <- data.frame( 
    tissue_name = tissue_name,
    # qc_metrics = qc_metrics,
    # mitochondrion = mitochondrion,
    discard = as.logical(mitochondrion$discard), 
    threshold = mitochondrion$threshold, 
    high_mitochondrion = mitochondrion$high_mitochondrion,
    subsets_Mito_sum = mitochondrion$subsets_Mito_sum,
    subsets_Mito_percent = mitochondrion$subsets_Mito_percent
  )
  return(plot_mito)
}

all_data <- lapply(seq_along(params$x1), function(i) {
  plot_mito_data(params$x1[[i]], sample_names[[i]], params$x3[[i]])
})

# Combine all data into a single tibble
combined_plot_mito_data <- bind_rows(all_data)

plot_each_sample <- function(combined_plot_mito_data) {
    # browser()
  num_tissues <- length(unique(combined_plot_mito_data$tissue_name))
  plot <- ggplot(combined_plot_mito_data, aes(x = subsets_Mito_sum, y = subsets_Mito_percent)) +
      facet_wrap(~ tissue_name) + 
      geom_point(aes(color = combined_plot_mito_data$high_mitochondrion), alpha = 0.5) +
      scale_x_log10() +
      geom_hline(yintercept = combined_plot_mito_data$threshold, color = "red", linetype = "dashed") +
      labs(x = "Total count", y = "Mitochondrial %", 
         title = paste("Percentage library size vs library size with", num_tissues, "tissue types"), 
         color = "High mitochondrial percentage") +
      theme_minimal() 
  
  #    unique_tissues <- unique(combined_plot_mito_data$tissue_name)
  #   for(tissue in unique_tissues) {
  #   tissue_data <- combined_plot_mito_data[combined_plot_mito_data$tissue_name == tissue,]
  #   threshold_value <- unique(tissue_data$threshold) # assuming there's one threshold per tissue
  #   plot <- plot + geom_hline(data = tissue_data, aes(yintercept = threshold_value), color = "red", linetype = "dashed")
  # 
  # }
  return(plot)
}
plot_each_sample(combined_plot_mito_data)

```


<!-- ```{r echo=FALSE, message=FALSE, warning=FALSE} -->
<!-- sample_names<- unlist(sample_names) -->
<!-- plot_data <- dplyr::tibble( -->
<!--   tissue_name = unique(sample_names), -->
<!--   input_files = params$x1, -->
<!--   empty_droplets = params$x2 -->
<!-- ) %>% -->
<!-- group_by(tissue_name) %>% -->
<!-- mutate(MA_plot = map2( -->
<!--   input_files, empty_droplets, -->
<!--   ~ { -->
<!--     # browser() -->
<!--       input_seurat = .x -->
<!--       empty_droplets = .y -->
<!--       input <- input_seurat@meta.data |> -->
<!--         tibble::rownames_to_column(var = '.cell') -->
<!--       joined_data <- empty_droplets |> -->
<!--         left_join(input |> dplyr::select(.cell, Tissue), by = '.cell') -->
<!--       mito_genes <- grep('^MT-', rownames(input_seurat[[assay]]), value=T) -->
<!--       ribo_genes <-grep('^RP(S|L)', rownames(input_seurat[[assay]]), value=T) -->
<!--       col <- rep('black',ncol(input_seurat)) -->
<!--       col[rownames(input_seurat[[assay]]) %in% mito_genes] <-'cyan' -->
<!--       col[rownames(input_seurat[[assay]]) %in% ribo_genes] <-'blueviolet' -->
<!--       col <- factor(col, levels = c('black', 'cyan', 'blueviolet'), labels = c('other', 'mito', 'ribo')) -->

<!--       extra <- joined_data |> dplyr::filter(FDR < 0.001) |> pull(.cell) -->
<!--       amb  <- joined_data |> dplyr::filter(is.na(FDR)) |> pull(.cell) -->
<!--       # sample_names <- joined_data$Tissue -->

<!--       # Sum of RNA-seq counts for each cell (or droplet) that is considered likely to be empty based on the FDR threshold -->
<!--       extra.mu <- input_seurat |> dplyr::filter(.cell %in% extra) |> GetAssayData(assay,slot = "counts") |> rowSums() -->
<!--       # Ambiguous empty droplets -->
<!--       amb.mu  <- input_seurat |> dplyr::filter(.cell %in% amb) |> GetAssayData(assay,slot = "counts") |> rowSums() -->

<!--       # ggplot_object <- edgeR::maPlot(extra.mu, amb.mu, normalize = TRUE, col = col, plot.it = FALSE) |> -->
<!--       # ggplot -->
<!--        # enframe() |> -->
<!--        #  pivot_wider() |> -->
<!--        #  unnest(A, M, w, v) |> -->
<!--        #  ggplot(aes(A, M, color = col, size = col)) + -->
<!--        #  geom_point() + -->
<!--        #  scale_color_manual(values = c("black", "red", "blue")) + -->
<!--        #  scale_size_manual(values = c(0.2, 2, 2)) + -->
<!--        #  theme_bw() -->
<!--     # # Create a data frame to return -->
<!--     data_to_plot <- data.frame( -->
<!--       A = extra.mu, -->
<!--       M = amb.mu, -->
<!--       color = col, -->
<!--       tissue = tissue_name -->
<!--     ) -->
<!--     data_to_plot -->
<!--   } -->
<!-- )) %>% -->
<!-- pull(MA_plot)|> -->

<!--   bind_rows() -->

<!-- # plot_data -->
<!-- ggplot(plot_data, aes(x = A, y = M, color = color)) + -->
<!--   geom_point(alpha = 0.5) + -->
<!--   facet_wrap(~ tissue) + -->
<!--   scale_color_manual(values = c("others" = "grey", "mito" = "cyan", "ribo" = "blueviolet")) + -->
<!--   labs(x = "True empty", y = "Ambiguous", color = "Content") + -->
<!--   theme_minimal() -->
<!-- ``` -->

